File: LpSolve/Elements/Line.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using LpSolve.Interface;
using MathExt;

namespace LpSolve.Elements
{
	public class Line : IElement<Line>
	{
		private Point _point;
		private Vector _vector;

		public Point Point { get { return this._point; } }
		public Vector Vector { get { return this._vector; } }

		public double C
		{
			get
			{
				return this._vector.X * this._point.X + this._vector.Y * this._point.Y;
			}
		}

		public Line(Point point, Vector vector)
		{
			if (point.GetDimension() != vector.GetDimension())
			{
				throw new ArgumentException(string.Format("Point and Vector ({{0}} and {{1}}) are of different dimensions!", point.GetDimension(), vector.GetDimension()));
			}

			this._point = point;
			this._vector = vector;
		}

		public Line MoveDown(Plane plane)
		{
			var point = this._point.MoveDown(plane, this._vector);
			var vec = this._vector.MoveDown(plane);
			return new Line(point, vec);
		}

		public int GetDimension()
		{
			return this._point.GetDimension();
		}

		public Point IntersectPlane(Plane plane)
		{
			var fakeLinePoint = this._point;
			var fakeLineVector = this._vector;

			if (plane.GetDimension() == 2)
			{
				//it is the line
				//|ax + by - c = 0
				//|a1x + b1y - c1 = 0

				var matrixA = Matrix.Create(new double[][] { 
					new double[] { this._vector.X, this._vector.Y },
					new double[] { plane.Vector.X, plane.Vector.Y}
				});

				var matrixB = Matrix.Create(new double[][] {
					new double[] { this.C },
					new double[] { plane.D}
				});

				var solver = new LESolver();

				var result = solver.Solve(matrixA, matrixB);

				return new Point(result, this._point.ParentPoint);
			}

			if (plane.GetDimension() == 3)
			{

			}

			throw new NotImplementedException("Implemented only for 2d");
		}
	}
}

File: cpp/Elements/Line.cpp
#include "Line.h"
#include "Point.h"
#include "Vector.h"
#include "Plane.h"

using namespace LpSolveCpp::Interface;
using namespace MathExt;
namespace LpSolveCpp
{
	namespace Elements
	{

		LpSolveCpp::Elements::Point *Line::getPoint() const
		{
			return this->_point;
		}

		LpSolveCpp::Elements::Vector *Line::getVector() const
		{
			return this->_vector;
		}

		double Line::getC() const
		{
			return this->_vector->getX() * this->_point->getX() + this->_vector->getY() * this->_point->getY();
		}

		Line::Line(LpSolveCpp::Elements::Point *point, LpSolveCpp::Elements::Vector *vector)
		{
			if (point->GetDimension() != vector->GetDimension())
			{
				throw std::invalid_argument("Point and Vector are of different dimensions!");
			}

			this->_point = point;
			this->_vector = vector;
		}

		Line *Line::MoveDown(Plane *plane)
		{
			auto point = this->_point->MoveDown(plane, this->_vector);
			auto vec = this->_vector->MoveDown(plane);
			return new Line(point, vec);
		}

		int Line::GetDimension()
		{
			return this->_point->GetDimension();
		}

		LpSolveCpp::Elements::Point *Line::IntersectPlane(Plane *plane)
		{
			auto fakeLinePoint = this->_point;
			auto fakeLineVector = this->_vector;

			if (plane->GetDimension() == 2)
			{
				//it is the line
				//|ax + by - c = 0
				//|a1x + b1y - c1 = 0

				std::vector<std::vector<double>> arrA;
				std::vector<double> arrA1;
				arrA1.push_back(this->_vector->getX());
				arrA1.push_back(this->_vector->getY());

				std::vector<double> arrA2;
				arrA2.push_back(plane->getVector()->getX());
				arrA2.push_back(plane->getVector()->getY());

				arrA.push_back(arrA1);
				arrA.push_back(arrA2);


				auto matrixA = Matrix::Create(arrA);

				auto matrixB = Matrix::Create(std::vector<std::vector<double>>
				{
					{this->getC()},
					{plane->getD()}
				});

				auto solver = new LESolver();

				auto result = solver->Solve(matrixA, matrixB);

				return new Point(result, this->_point->getParentPoint());
			}

			if (plane->GetDimension() == 3)
			{

			}

			throw NotImplementedException("Implemented only for 2d");
		}
	}
}
